src/TaggableCachePoolTest.php

```diff
@@ -292,6 +275,20 @@ public function testInvalidateTags()
         $this->cache->save($item);
         $this->cache->invalidateTags(['tag1']);
 
-        $this->assertTrue($this->cache->hasItem('key'), 'Item key list should be removed when clearing the tags');
+        $this->assertTrue($this->cache->hasItem('key'), 'Item k list should be removed when clearing the tags');
+    }
+
+    /**
+     * When an item is overwritten we need to clear tags for original item.
+     */
+    public function testTagsAreCleanedOnSave()
+    {
+        $pool = $this->cache;
+        $i    = $pool->getItem('key');
+        $pool->save($i->addTag('foo'));
+        $i = $pool->getItem('key');
+        $pool->save($i->addTag('bar'));
+        $pool->invalidateTags(['foo']);
+        $this->assertTrue($pool->getItem('key')->isHit());
```

Comment:
IMHO, this is expected, see https://github.com/symfony/symfony/issues/19728#issuecomment-267920552 for a description of `getTags`' behavior.
Tags are an artifact of the *computation* of the cached data. Adding a tag without actually computing the value makes no sense in this respect.